imap_processing/idex/idex_l2a.py

```diff
@@ -472,3 +497,134 @@ def calculate_area_under_emg(time_slice: np.ndarray, param: np.ndarray) -> float
+
+    Returns
+    -------
+    param : numpy.ndarray
+        Optimal target signal fit values for the parameters (popt)
+        [time_of_impact, constant_offset, amplitude, rise_time, discharge_time]
+        if fit successful. None otherwise.
+    sig_amp : float
+        Signal amplitude, calculated as difference between fitted maximum signal
+        and baseline mean if fit successful. None otherwise.
+    chi_squared : float
+        Sum of squared residuals from the fit.
+    reduced_chi_squared : float
+        Chi-squared per degree of freedom.
+    result : numpy.ndarray
+        The model values evaluated at each time point.
+    """
+    # TODO: The IDEX team is iterating on this Function and will provide more
+    #         information soon.
+    signal = np.array(target_signal.data)
+    time = np.array(low_sampling_time.data)
+    if remove_noise:
+        pass
+    # Time before image charge
+    pre = -2.0
+    # Get signal values where the time is before the image charge
+    signal_before_imapact = signal[time < pre]
+    # Center the baseline signal around zero
+    signal_baseline = signal_before_imapact - np.mean(signal_before_imapact)
+
+    # Initial Guess for the parameters of the ion grid signal
+    time_of_impact = 0.0  # Time of dust hit
+    constant_offset = 0.0  # Initial baseline
+    amplitude: float = np.max(signal)  # Signal height
+    rise_time = 0.371  # How fast the signal rises (s)
+    discharge_time = 0.371  # How fast signal decays (s)
+
+    p0 = [time_of_impact, constant_offset, amplitude, rise_time, discharge_time]
+
+    try:
+        param, _ = curve_fit(
+            fit_impact,
+            time,
+            signal,
+            p0=p0,
+            maxfev=100_000,  # , epsfcn=1e-10
+        )
+    except RuntimeError as e:
+        logger.warning(
+            f"Failed to fit curve: {e}\n"
+            f"Time range: {time[0]:.2f} to {time[-1]:.2f}\n"
+            f"Signal range: {min(signal):.2f} to {max(signal):.2f}\n"
+            "Returning None."
+        )
+        return (
+            np.full(len(p0), np.nan),
+            np.nan,
+            np.nan,
+            np.nan,
+            np.full_like(time, np.nan),
+        )
+
+    impact_fit = fit_impact(time, *param)
+    # Calculate the resulting signal amplitude after removing baseline noise
+    sig_amp = max(impact_fit) - np.mean(signal_baseline)
+
+    # Calculate chi square and reduced chi square
+    chisqr = float(np.sum((signal - impact_fit) ** 2))
+    # To get reduced chi square divide by dof (number of points - number of params)
+    redchi = chisqr / (len(signal) - len(p0))
+
+    return param, float(sig_amp), chisqr, redchi, impact_fit
+
+
+def fit_impact(
+    time: np.ndarray,
+    time_of_impact: float,
+    constant_offset: float,
+    amplitude: float,
+    rise_time: float,
+    discharge_time: float,
+) -> NDArray:
+    """
+    Fit function for the Ion Grid and two target signals given by (Horanyi, 2014).
```

Comment:
Nice reference! We should include the actual paper citation here too.

https://numpydoc.readthedocs.io/en/latest/format.html#references